### PR TITLE
Custom errors to display when image load fails

### PR DESCRIPTION
**Pull request by wuweiweiwu**
#82 

@fritz-c Should I change the default error message to be a dom element instead of string? 

Provides more customizability.

### CODE REVIEW

File: src/react-image-lightbox.js [hunk 1]
```diff
@@ -1180,6 +1191,15 @@ class ReactImageLightbox extends Component {
     this.getSrcTypes().forEach(srcType => {
       const type = srcType.name;
 
+      // there is no error when we try to load it initially
+      if (props[type]) {
+        this.setState(prevState => ({
+          loadErrorStatus: Object.assign({}, prevState.loadErrorStatus, {
```
Comment (fritz-c) [inline src/react-image-lightbox.js]: Please use the object rest spread operator, like so:
```js
{ ...prevState.loadErrorStatus, [type]: false }
```
rather than `Object.assign`. This is partially about code style, partially about browser compatibility (Babel transpiles the spread operator code to ES5)

File: src/react-image-lightbox.js [hunk 2]
```diff
@@ -1313,7 +1339,25 @@ class ReactImageLightbox extends Component {
         imageStyle.cursor = 'move';
       }
 
-      if (bestImageInfo === null) {
+      // when error on one of the loads then push custom error stuff
+      if (
+        bestImageInfo === null &&
+        Object.values(loadErrorStatus).includes(true)
```
Comment (fritz-c) [inline src/react-image-lightbox.js]: Pretty sure `Object.values` and `includes` might be iffy when it comes to browser compatibility. At worst I want to support IE 11, with barebones / fatal-error-free functionality in IE 9.

File: src/style.scss [hunk 3]
```diff
@@ -297,6 +297,30 @@ $maxCaptionHeight: 150px;
   right: 0;
   bottom: 0;
   left: 0;
+  .imagePrev & {
+    display: none;
```
Comment (fritz-c) [inline src/style.scss]: Does this not cause issues with animation when the adjacent images load faster?

File: src/react-image-lightbox.js [hunk 4]
```diff
@@ -1142,6 +1145,14 @@ class ReactImageLightbox extends Component {
 
     inMemoryImage.onerror = errorEvent => {
       this.props.onImageLoadError(imageSrc, srcType, errorEvent);
+
+      // failed to load so set the state loadErrorStatus
+      this.setState(prevState => ({
+        loadErrorStatus: Object.assign({}, prevState.loadErrorStatus, {
```
Comment (fritz-c) [inline src/react-image-lightbox.js]: Please use the object rest spread operator here (see other comment for details).

File: src/react-image-lightbox.js [hunk 5]
```diff
@@ -1759,6 +1803,8 @@ ReactImageLightbox.propTypes = {
   zoomInLabel: PropTypes.string,
   zoomOutLabel: PropTypes.string,
   closeLabel: PropTypes.string,
+
+  imageLoadErrorMessage: PropTypes.string
```
Comment (fritz-c) [inline src/react-image-lightbox.js]: I think `PropTypes.node` would be more versatile. That way they could plug in any element they wanted to in that area.

File: src/react-image-lightbox.js (hunk 1 above)
Comment (wuweiweiwu) [reply src/react-image-lightbox.js]: sounds good!

File: src/react-image-lightbox.js (hunk 2 above)
Comment (wuweiweiwu) [reply src/react-image-lightbox.js]: I will find a compatible solution.

File: src/style.scss (hunk 3 above)
Comment (wuweiweiwu) [reply src/style.scss]: I do this so not to display the loading and/or error message on the previous or next images on the current image

**Comment by wuweiweiwu** [comment]
Seems like coveralls is down. From the travis build

```
Bad response: 503 <h2>This website is under heavy load (queue full)</h2><p>We're sorry, too many people are accessing this website at the same time. We're working on this problem. Please try again later.</p>
```

File: src/react-image-lightbox.js [hunk 6]
```diff
@@ -1313,7 +1335,36 @@ class ReactImageLightbox extends Component {
...
+      // support IE 9 and 11
+      const hasTrueValue = object => {
+        const keys = Object.keys(object);
+        for (let i = 0; i < keys.length; i += 1) {
+          if (object[keys[i]]){
+            return true;
+          }
+        }
+        return false;
+      }
```
Comment (fritz-c) [inline src/react-image-lightbox.js]: Could also shorten this to
```js
Object.keys(object).some(key => object[key])
```

File: src/react-image-lightbox.js [hunk 7]
```diff
@@ -1313,7 +1335,36 @@ class ReactImageLightbox extends Component {
...
+
+      // when error on one of the loads then push custom error stuff
+      if (
+        bestImageInfo === null &&
+        hasTrueValue(loadErrorStatus)
```
Comment (fritz-c) [inline src/react-image-lightbox.js]: Could you fix the code formatting with `npm run prettier` before I merge this?

File: src/react-image-lightbox.js [hunk 8]
```diff
@@ -1180,6 +1189,13 @@ class ReactImageLightbox extends Component {
     this.getSrcTypes().forEach(srcType => {
       const type = srcType.name;
 
+      // there is no error when we try to load it initially
+      if (props[type]) {
```
Comment (fritz-c) [inline src/react-image-lightbox.js]: Maybe add another condition to check if the `this.state.loadErrorStatus[type]` is `true` before setting it to `false` here. I want to avoid unnecessary rerenders due to `this.setState`

File: src/style.scss (hunk 3 above)
Comment (fritz-c) [reply src/style.scss]: Aaah, so that's the source of that bug. Thanks!

File: src/react-image-lightbox.js [hunk 9]
```diff
@@ -1313,7 +1335,36 @@ class ReactImageLightbox extends Component {
...
+      }
+
+      // when error on one of the loads then push custom error stuff
+      if (
+        bestImageInfo === null &&
+        hasTrueValue(loadErrorStatus)
+      ) {
+        images.push(
+          <div
+            className={`${imageClass} ${styles.image} ril-errored `}
```
Comment (fritz-c) [inline src/react-image-lightbox.js]: Got some extra whitespace on the end, there.

File: src/__tests__/react-image-lightbox.spec.js [hunk 10]
```diff
@@ -156,3 +156,34 @@ describe('Snapshot Testing', () => {
...
+    wrapper.setState({
+      loadErrorStatus: { mainSrc : true }
+    });
+    wrapper.setProps({
+      imageLoadErrorMessage
+    });
+    wrapper.update();
+    expect(wrapper.find('div.errorContainer')).toContainReact(imageLoadErrorMessage);
+  });
+});
```
Comment (fritz-c) [inline src/__tests__/react-image-lightbox.spec.js]: Very nice. Thanks!

File: src/style.scss [hunk 11]
```diff
@@ -297,6 +297,30 @@ $maxCaptionHeight: 150px;
...
+.errorContainer {
+  position: absolute;
+  top: 0;
+  right: 0;
+  bottom: 0;
+  left: 0;
+  display: flex;
+  align-items: center;
+  justify-content: center;
+  color: #ff0000;
```
Comment (fritz-c) [inline src/style.scss]: I think `#fff` would be a good color in this context. Bright red is a bit too alarming, and I think the fact that text comes up instead of an image is warning enough that something went wrong.

**Comment by fritz-c** [comment]
Roger that on the Coveralls situation. Worst case, I'll run it on my own machine.

**Comment by fritz-c** [comment]
Looks good! I made you a collaborator on this repo, so see if you can merge it yourself. One thing about features / fixes, though: I'm using [standard-version](https://github.com/conventional-changelog/standard-version) for updating my changelog, which parses commits that match the format `feat: <message about what feature was added>` (as well as `fix:` for fixes, etc.), and automatically generates the changelog. So when you merge, change the message to meet that format. In this case, you've got a feature that also fixes something, so just use `feat:`, which will bump the minor version number, and maybe a note in the description like `fixes #82`.

**Comment by wuweiweiwu** [comment]
@fritz-c Should I squash it as well?

**Comment by fritz-c** [comment]
Excellent!
